src/java.base/share/classes/java/util/Locale.java

```diff
@@ -2257,7 +2260,7 @@ public boolean equals(Object obj) {
     /**
      * Calculated hashcode
      */
-    private transient volatile int hashCodeValue;
```

Comment:
We can additionally annotate such cache fields with `@Stable` so as to enable constant-folding optimizations from the hotspot JIT.

Reply:
These fields can only be written once besides the default values, but they don't have to be written in the static initializer or constructor. So when a non-zero hash code is written, it's cached as if it's a final field. For a zero value, it would always undergo extra calculations like before and write multiple times, but the writes don't matter as it would only write zero, which would not yield false hash code if a previously written 0 value is cached.

Reply:
> Shouldn't the fields annotated with `@Stable` be `final` as well?

You might have seen `@Stable final` fields. Those are only meaningful for arrays, where such caching happens at each array index. For non-arrays, `@Stable final` is simply equivalent to `final`, and simple `@Stable` acts as what I said above.